File: primalTestDotNet/AI/States/StateBase.cs
using PrimalTestDotNet.Entities;

namespace PrimalTestDotNet.AI.States;

public class DistanceData
{
    public int ShortestDistance { get; set; } = int.MaxValue;
    public IntVector2? PreviousVertex { get; set; }
}

/// <summary>
/// Base class for AI states. Contains a crude implementation of Dijkstra's algorithm for path calculations
/// </summary>
public abstract class StateBase(Level.Level level)
{
    protected Level.Level _level = level;

    /// <summary>
    /// Returns a list of coordinates of the shortest path possible between two given coordinates.
    /// </summary>
    /// <param name="from">The start coordinate</param>
    /// <param name="to">The final destination</param>
    /// <param name="distanceData">A map of values calculated by <see cref="CalculateDistanceData(IntVector2, IEnumerable{IGameObject})"/></param>
    /// <returns></returns>
    protected List<IntVector2> GetPath(IntVector2 from, IntVector2 to, Dictionary<IntVector2, DistanceData> distanceData)
    {
        var path = new List<IntVector2>
        {
            to
        };

        IntVector2? node = to;
        while (node != null)
        {
            var prev = distanceData[node.Value].PreviousVertex;
            if (prev != null && prev.Value != from)
                path.Add(prev.Value);
            node = prev;
        }

        return path;
    }

    /// <summary>
    /// Calculates shortest paths from all nodes to the start node for the <see cref="Level.Level"/> data with Dijkstra's algorithm
    /// </summary>
    /// <param name="start"></param>
    /// <param name="additionalWalls">Additional walls other than the one provided by <see cref="Level.Level"/></param>
    /// <returns></returns>
    protected Dictionary<IntVector2, DistanceData> CalculateDistanceData(IntVector2 start, IEnumerable<IGameObject> additionalWalls)
    {
        var distanceData = new Dictionary<IntVector2, DistanceData>();

        var visited = new List<IntVector2>();
        var unvisited = new List<IntVector2>();
        for (int x = 0; x < _level.CollisionData.GetLength(1); x++)
        {
            for (int y = 0; y < _level.CollisionData.GetLength(0); y++)
            {
                if (_level.CollisionData[y, x] == false && !additionalWalls.Any(o => o.Position == new IntVector2(x, y)))
                    unvisited.Add(new IntVector2(x, y));
            }
        }

        unvisited.ForEach(x =>
        {
            distanceData[x] = new DistanceData
            {
                PreviousVertex = null,
                ShortestDistance = x == start ? 0 : int.MaxValue
            };
        });


        while (unvisited.Count > 0
            && unvisited.Any(u => distanceData[u].ShortestDistance < int.MaxValue)) // disregard isolated nodes
        {
            var node = unvisited.OrderBy(n => distanceData[n].ShortestDistance).First();
            unvisited.Remove(node);

            var neighbours = GetNeighbours(node, unvisited);
            foreach (var neighbour in neighbours)
            {
                var distance = distanceData[node].ShortestDistance + 1;
                if (distance < distanceData[neighbour].ShortestDistance)
                {
                    distanceData[neighbour].ShortestDistance = distance;
                    distanceData[neighbour].PreviousVertex = node;
                }
            }

            visited.Add(node);
        }

        return distanceData;
    }

    private static IEnumerable<IntVector2> GetNeighbours(IntVector2 node, List<IntVector2> unvisited)
    {
        var possibleMoves = new IntVector2[]
        {
            new(0, 1),  // Up
            new(0, -1), // Down
            new(1, 0),  // Right
            new(-1, 0)  // Left
        };

        return possibleMoves.Select(move => new IntVector2(node.X + move.X, node.Y + move.Y)).Where(unvisited.Contains);
    }
}

File: PrimalTestCpp/StateBase.cpp
#include "StateBase.h"
#include "Level.h"
#include "IDrawable.h"
#include <limits>
#include "IntVector2D.h"
#include "Hero.h"
#include <optional>

StateBase::StateBase(Level& level) : _level(level)
{
}

std::unordered_map<IntVector2D, DistanceData> StateBase::calculateDistanceData(const IntVector2D& start, const std::vector<std::shared_ptr<IGameObject>>& additionalWalls)
{
	std::unordered_map<IntVector2D, DistanceData> distanceData;

	std::vector<IntVector2D> visited;
	std::vector<IntVector2D> unvisited;
	for (unsigned int x = 0; x < _level.getX(); x++) {
		for (unsigned int y = 0; y < _level.getY(); y++) {
			if (_level.getCollisionData()[y][x] == false) {
				IntVector2D pos(x, y);
				bool isWall = std::any_of(additionalWalls.begin(), additionalWalls.end(),
					[&pos](const std::shared_ptr<IGameObject>& obj) {
						return obj->getPosition() == pos;
					});

				if (!isWall) {
					unvisited.push_back(pos);
				}
			}
		}
	}

	for (const auto& node : unvisited) {
		DistanceData d;
		d.previousVertex = std::nullopt;
		d.shortestDistance = (node == start) ? 0 : INT_MAX;
		distanceData[node] = d;
	}

	while (!unvisited.empty()
		&& std::any_of(unvisited.begin(), unvisited.end(),
			[&distanceData](const IntVector2D& vec) {
				return distanceData[vec].shortestDistance < INT_MAX;
			})) {

		auto nodeIt = std::min_element(unvisited.begin(), unvisited.end(),
			[&distanceData](const IntVector2D& a, const IntVector2D& b) {
				return distanceData[a].shortestDistance < distanceData[b].shortestDistance;
			});
		IntVector2D node = *nodeIt;
		unvisited.erase(nodeIt);

		auto neighbors = getNeighbors(node, unvisited);
		for (const auto& neighbor : neighbors) {
			int distance = distanceData[node].shortestDistance + 1; // Every neighbor is 1 away
			if (distance < distanceData[neighbor].shortestDistance) {
				// if the calculated distance is shorter than a previously calculated, update it
				distanceData[neighbor].shortestDistance = distance;
				distanceData[neighbor].previousVertex = node;
			}
		}

		visited.push_back(node);
	}

	return distanceData;
}

std::vector<IntVector2D> StateBase::getNeighbors(const IntVector2D& node, std::vector<IntVector2D> unvisited)
{
	std::vector<IntVector2D> possibleMoves = {
		{0, 1},  // Up
		{0, -1}, // Down
		{1, 0},  // Right
		{-1, 0}  // Left
	};

	std::vector<IntVector2D> neighbors;
	for (auto& move : possibleMoves) {
		IntVector2D neighbour(node.getX() + move.getX(), node.getY() + move.getY());
		if (std::find(unvisited.begin(), unvisited.end(), neighbour) != unvisited.end()) {
			neighbors.push_back(neighbour);
		}
	}

	return neighbors;
}

std::vector<IntVector2D> StateBase::getPath(const IntVector2D& from, const IntVector2D& to,
	const std::unordered_map<IntVector2D, DistanceData>& distanceData) {
	std::vector<IntVector2D> path = { to };

	std::optional<IntVector2D> node = to;

	while (node.has_value()) {
		auto it = distanceData.find(node.value());
		if (it != distanceData.end()) {
			auto prev = it->second.previousVertex;
			if (prev.has_value() && prev.value() != from) {
				path.push_back(prev.value());
			}
			node = prev;
		}
		else {
			node.reset();
		}
	}

	return path;
}
